### PR TITLE
K nn algorithm

### PR DESCRIPTION
**Pull request by aubrey-m-merritt**
testing doesn't have the comparison to the accurate value yet since we're testing the functional
Set up of KNN Model using tribou

### CODE REVIEW

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                else {
+                    input = SparseVector.createSparseVector(example, featureIDMap, false);
+                }
+                // nothing is found
+                if (input.numActiveElements() == 0) {
+                    throw new IllegalArgumentException("No features found in Example " + example);
+                }
+            }
+            catch(Exception e){
+                System.out.print("Error at Prediction<T> predict:" + e);
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:194]: Use the logger

File: src/main/java/com/github/gissuite/gribinterpolation/data/DataPoint.java [hunk 2]
```diff
@@ -6,6 +6,7 @@ public class DataPoint {
     private float latitude;
     private float temperatureK;
     private float depth;
+    private float distance;
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/data/DataPoint.java:9]: Distance shouldn't be stored on DataPoint

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+public class NearestNeighbor {
+    /**
+     * @param dataPoints All the DataPoints in the data set
+     * @param longitudeToInterpolate The longitude value of the point needed to be interpolated
+     * @param latitudeToInterpolate  The latitude value of the point needed to be interpolated
+     * @param k How many nearest neighbors to return
+     * @param amountOfDataPoints The total about of DataPoints
+     * @return ArrayList of nearest neighbors (DataPoints)
+     */
+        public static ArrayList<DataPoint> getNearestNeighbor(ArrayList<DataPoint> dataPoints, float longitudeToInterpolate, float latitudeToInterpolate, int k, int amountOfDataPoints) {
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:47]: Please make parameter names less verbose

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            HashMap<DataPoint, Double> hashMap = new HashMap<>();
+            for (int i = 0; i < amountOfDataPoints; i++) {
+                double distance = DistanceFinder.haverSine(dataPoints.get(i).getLatitude(), dataPoints.get(i).getLongitude(), latitudeToInterpolate, longitudeToInterpolate);
+                hashMap.put(dataPoints.get(i), distance);
+            }
+            Map<DataPoint, Double> sortedHashMap = sortByValue(hashMap);
+            Set<DataPoint> keySet = sortedHashMap.keySet();
+            Iterator<DataPoint> iterator = keySet.iterator();
+            int count = 0;
+            while (iterator.hasNext() && count < k) {
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:58]: `iterator.hasNext()` should suffice

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                // get the distance
+                distance = DistanceFinder.haverSine(nearestNeighbor.getLatitude(), nearestNeighbor.getLongitude(), toInterpolate.getLatitude(), toInterpolate.getLongitude());
+                // getting temperature over distance
+                tempOverDist = (nearestNeighbor.getTemperatureK()) / distance;
+                // getting total temperature over distance
+                totalTempOverDist += tempOverDist; // the numerator of our formula
+
+
+                if(Double.isFinite(distance)){
+                    // need to check for when distance is zero because I was getting infinity (not good)
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:101]: Why?

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                        totalDistanceInverse += 0; // probably need to change this
+                    }
+                }
+                else{
+                    throw new ArithmeticException("Distance is infinity");
+                }
+                totalDistance += distance; // store the distance
+            }
+            num = totalTempOverDist/totalDistanceInverse;
+            interpolatedTemp = Double.parseDouble(df.format(num));
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:116]: You shouldn't need to parse or format this

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        }
+        catch(Exception e){
+            System.out.println("Problem with knnInterpolation: " + e);
+            System.out.println("String being parsed: " + num); // Print the string being parsed
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:120]: Use the Logger

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            this.neighboursQueryFactory = neighboursQueryFactory;
+            this.neighboursQuery = neighboursQueryFactory.createNeighboursQuery(getSGDVectorArr());
+        }
+
+        /**
+         *
+         */
+        private SGDVector[] getSGDVectorArr() { // Converting Pair<SGDVector> to SGDVector[]
+            SGDVector[] sgdVectors = new SGDVector[vectors.length];
+            int n = 0;
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:162]: What is the purpose of `n`?

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                if (example.size() == featureIDMap.size()) {
+                    input = DenseVector.createDenseVector(example, featureIDMap, false);
+                }
+                // Something was found, but example doesn't match
+                else {
+                    input = SparseVector.createSparseVector(example, featureIDMap, false);
+                }
+                // nothing is found
+                if (input.numActiveElements() == 0) {
+                    throw new IllegalArgumentException("No features found in Example " + example);
```
Comment (sechols223) [inline src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:190]: Don't throw an exception, just log it

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+                float latitudeToInterpolate = 2;
+                float longitudeToInterpolate = 3;
+                // Number of nearest neighbors to find
+                int k = 3;
+                // Call the method to get nearest neighbors
+                ArrayList<DataPoint> nearestNeighbors = NearestNeighbor.getNearestNeighbor(neighbors, longitudeToInterpolate, latitudeToInterpolate, k, neighbors.size());
+                // Assert the result
+                System.out.println("Nearest Neighbors:");
+                for (DataPoint neighbor : nearestNeighbors) {
+                    System.out.println("Longitude: " + neighbor.getLongitude() + ", Latitude: " + neighbor.getLatitude() + ", Temperature: " + neighbor.getTemperatureK() + ", Depth: " + neighbor.getDepth());
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:42]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                }
+            }
+            catch(Exception e){
+                System.out.println("Problem with getNearestNeighbor:" + e);
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:46]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                int k = 3;
+                // Call the method to get nearest neighbors
+                ArrayList<DataPoint> nearestNeighbors = NearestNeighbor.getNearestNeighbor(neighbors, longitudeToInterpolate, latitudeToInterpolate, k, neighbors.size());
+                // Assert the result
+                System.out.println("Nearest Neighbors:");
+                for (DataPoint neighbor : nearestNeighbors) {
+                    System.out.println("Longitude: " + neighbor.getLongitude() + ", Latitude: " + neighbor.getLatitude() + ", Temperature: " + neighbor.getTemperatureK() + ", Depth: " + neighbor.getDepth());
+                }
+            }
+            catch(Exception e){
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:45]: Don't catch exceptions in unit tests

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+            try{
+                // Interpolation point coordinates
+                float latitudeToInterpolate = 2;
+                float longitudeToInterpolate = 3;
+                // Number of nearest neighbors to find
+                int k = 3;
+                // Call the method to get nearest neighbors
+                ArrayList<DataPoint> nearestNeighbors = NearestNeighbor.getNearestNeighbor(neighbors, longitudeToInterpolate, latitudeToInterpolate, k, neighbors.size());
+                // Assert the result
+                System.out.println("Nearest Neighbors:");
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:40]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                ArrayList<DataPoint> nearestNeighbors = getNearestNeighbor(neighbors, interpolatedPoint.getLongitude(), interpolatedPoint.getLatitude(),  k, neighbors.size());
+
+                // KNN interpolation for temperature
+                double interpolatedTemp = knnInterpolation(nearestNeighbors, interpolatedPoint);
+
+                // Setting the temperature
+                interpolatedPoint.setTemperatureK((float) interpolatedTemp);
+
+                // Showing the new temperature
+                System.out.println("Interpolated Temp: " + interpolatedPoint.getTemperatureK());
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:61]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            catch (Exception e) {
+                System.out.println("Problem with knnInterpolation:" + e);
+            }
+        }
+        catch(Exception e){
+            System.out.println("Error with the overall :( : " + e);
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:68]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                double interpolatedTemp = knnInterpolation(nearestNeighbors, interpolatedPoint);
+
+                // Setting the temperature
+                interpolatedPoint.setTemperatureK((float) interpolatedTemp);
+
+                // Showing the new temperature
+                System.out.println("Interpolated Temp: " + interpolatedPoint.getTemperatureK());
+            }
+            catch (Exception e) {
+                System.out.println("Problem with knnInterpolation:" + e);
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:64]: Don't print

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+import static com.github.gissuite.gribinterpolation.core.NearestNeighbor.knnInterpolation;
+import static java.lang.Float.NaN;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+public class NearestNeighborTest {
+    @Test
+    public void knn_Should_Return_DataPoint(){
```
Comment (sechols223) [inline src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:17]: There aren't any assertions in this test

**Review by sechols223 — CHANGES_REQUESTED** [review]

**Review by sechols223 — CHANGES_REQUESTED** [review]
Changes need to be made

File: src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java (hunk 5 above)
Comment (aubrey-m-merritt) [reply src/main/java/com/github/gissuite/gribinterpolation/core/NearestNeighbor.java:101]: I was trying to fix the infinity problem.

File: src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java (hunk 13 above)
Comment (aubrey-m-merritt) [reply src/test/java/com/github/gissuite/gribinterpolation/NearestNeighborTest.java:40]: I was printing in this file mostly to check if things were working at every step.

**Pull request merged**